src/libpmemfile/preload.c

```diff
@@ -2908,6 +2909,72 @@ establish_mount_points(const char *config, struct stat *kernel_cwd_stat)
 	} while (config != NULL);
 }
 
+static int (*libc__xpg_strerror_r)(int __errnum, char *__buf, size_t __buflen);
+static char *(*libc_strerror_r)(int __errnum, char *__buf, size_t __buflen);
+static char *(*libc_strerror)(int __errnum);
+
+int __xpg_strerror_r(int __errnum, char *__buf, size_t __buflen);
+
+/*
+ * XSI-compliant version of strerror_r. We have to override it to handle
+ * possible deadlock/infinite recursion when pmemfile is called from inside of
+ * strerror_r implementation and we call back into libc because of some failure
+ * (notably: pool opening failed when process switching is enabled).
+ */
+int
+__xpg_strerror_r(int __errnum, char *__buf, size_t __buflen)
+{
+	if (!guard_flag && libc__xpg_strerror_r)
+		return libc__xpg_strerror_r(__errnum, __buf, __buflen);
+
+	if (__errnum == EAGAIN) {
+		const char *str =
+			"Resource temporary unavailable (pmemfile wrapper)";
+		if (__buflen < strlen(str) + 1)
+			return ERANGE;
+		strcpy(__buf, str);
+		return 0;
+	}
+
+	const char *str = "Error code %d (pmemfile wrapper)";
+	if (__buflen < strlen(str) + 10)
+		return ERANGE;
+	sprintf(__buf, str, __errnum);
+
+	return 0;
+}
+
+/*
+ * GNU-compliant version of strerror_r. See __xpg_strerror_r description.
+ */
+char *
+strerror_r(int __errnum, char *__buf, size_t __buflen)
+{
+	if (!guard_flag && libc_strerror_r)
+		return libc_strerror_r(__errnum, __buf, __buflen);
+
+	const char *str = "Error code %d (pmemfile wrapper)";
+	if (__buflen < strlen(str) + 10)
+		return NULL;
+
+	sprintf(__buf, str, __errnum);
```

Comment:
Are you sure calling `sprintf` is OK here?

Reply:
I'm not, alternatives?